File: statter/parsers/flexbar_fastq_parser.py
import gzip
import logging
from pathlib import Path
from typing import List, TextIO

from pyfastx import Fastx

logger = logging.getLogger(__name__)


class HeaderFixer:
    """
    Fix UMI positions in FASTQ headers produced by Flexbar.

    Flexbar appends UMIs to the end of the headers, which means that in Fastq files from Casava >1.8 versions
    the UMI is appended to the end of the comment section. All aligners typically remove this part of the header
    during alignment, which results in loss of UMI information.

    This class moves the UMI to the enf of the actual read identifier, before the first space separator.
    """

    def __init__(self, in_fq: str, out_fq: str, separator: str = "_") -> None:
        self.in_fq = in_fq
        self.out_fq = out_fq
        self.separator = separator

    def _writer(self) -> TextIO:
        out: Path = Path(self.out_fq)
        if out.suffix in set([".gz", ".gzip"]):
            return gzip.open(self.out_fq, "wt")
        else:
            return open(self.out_fq, "w")

    def fix_umi_header(self) -> None:
        """
        Process the input FASTQ file and write the modified headers to the output FASTQ file.
        """
        warnings: List[str] = []
        n_warnigs: int = 0
        buffer: List[str] = []
        buf_size: int = 50000
        with self._writer() as _fh:
            for name, seq, qual, comment in Fastx(self.in_fq, comment=True):
                if comment is None:
                    if len(warnings) < 10:
                        warnings.append("Detected header without comment section.")
                        n_warnigs += 1
                    buffer.append(f"@{name}\n{seq}\n+\n{qual}\n")
                    continue
                dat: List[str] = comment.split(self.separator)
                if len(dat) < 2:
                    if len(warnings) < 10:
                        warnings.append(
                            f"Header '{name}' does not contain expected UMI separator '{self.separator}'"
                        )
                        n_warnigs += 1
                    buffer.append(f"@{name} {comment}\n{seq}\n+\n{qual}\n")
                    continue
                umi: str = dat[-1]
                new_header: str = f"{name}{self.separator}{umi}"
                buffer.append(f"@{new_header} {' '.join(dat[:-1])}\n{seq}\n+\n{qual}\n")
                if len(buffer) >= buf_size:
                    _fh.writelines(buffer)
                    buffer.clear()
            if buffer:
                _fh.writelines(buffer)
        if warnings:
            for w in warnings:
                logger.warning(w)
            if n_warnigs > len(warnings):
                logger.warning(f"... and {n_warnigs - len(warnings)} more warnings.")

File: src/statter_rs/flexbar_fastq_parser.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader, BufWriter, Write};
use std::path::Path;
use flate2::read::GzDecoder;
use flate2::write::GzEncoder;
use flate2::Compression;

/// Fix UMI positions in FASTQ headers produced by Flexbar.
/// Moves the UMI from the comment section to the end of the read identifier.
pub fn fix_umi_header<P: AsRef<Path>>(in_fq: P, out_fq: P, separator: &str) -> io::Result<()> {
    let in_path = in_fq.as_ref();
    let out_path = out_fq.as_ref();
    let reader: Box<dyn BufRead> = if in_path.extension().map_or(false, |e| e == "gz" || e == "gzip") {
        Box::new(BufReader::new(GzDecoder::new(File::open(in_path)?)))
    } else {
        Box::new(BufReader::new(File::open(in_path)?))
    };
    let writer: Box<dyn Write> = if out_path.extension().map_or(false, |e| e == "gz" || e == "gzip") {
        Box::new(BufWriter::new(GzEncoder::new(File::create(out_path)?, Compression::default())))
    } else {
        Box::new(BufWriter::new(File::create(out_path)?))
    };
    let mut writer = BufWriter::new(writer);
    let mut lines = reader.lines();
    while let (Some(Ok(header)), Some(Ok(seq)), Some(Ok(plus)), Some(Ok(qual))) =
        (lines.next(), lines.next(), lines.next(), lines.next())
    {
        if !header.starts_with('@') {
            writeln!(writer, "{}", header)?;
            writeln!(writer, "{}", seq)?;
            writeln!(writer, "{}", plus)?;
            writeln!(writer, "{}", qual)?;
            continue;
        }
        let mut parts = header[1..].splitn(2, ' ');
        let name = parts.next().unwrap_or("");
        let comment = parts.next();
        if let Some(comment) = comment {
            let dat: Vec<&str> = comment.split(separator).collect();
            if dat.len() < 2 {
                writeln!(writer, "{}", header)?;
            } else {
                let umi = dat.last().unwrap();
                let new_header = format!(@"{}{}{} {}", name, separator, umi, dat[..dat.len()-1].join(" "));
                writeln!(writer, "@{}", new_header)?;
            }
        } else {
            writeln!(writer, "@{}", name)?;
        }
        writeln!(writer, "{}", seq)?;
        writeln!(writer, "{}", plus)?;
        writeln!(writer, "{}", qual)?;
    }
    Ok(())
}
